### PR TITLE
JS Driver: Add support for new exception types

### PR DESCRIPTION
**Pull request by marshall007**
See #4472. Includes updates to all polyglot tests, so testing the updates to Python and Ruby drivers should be easier now.

@mlucy there were no existing tests that returned an `RqlOpIndeterminateError`. I'm not sure if that's an oversight on my part or if there simply aren't any existing cases that cover that.

I was planning on running the Ruby tests to make sure all the changes to the polyglot tests line up, but I'm getting:

```
$ ./test/rql_test/test-runner -i rb
Usage: test-runner [options] [patterns]

test-runner: error: Unable to find a valid interpreter for language: rb
```

Also there are a few failing connections tests in the JS driver right now. Two of them are timeouts, and the others are "could not connect to server" errors for `r.http`. So there may be a couple tests in there that still need to be updated, but I was unsure of why they were failing to begin with.

### CODE REVIEW

**Comment by mlucy** [comment]
Awesome @marshall007 !

> @mlucy there were no existing tests that returned an RqlOpIndeterminateError. I'm not sure if that's an oversight on my part or if there simply aren't any existing cases that cover that.

That's correct, `OpIndeterminate` is only produced by events that cause internal disconnects in the cluster which the polyglot tests aren't set up to produce right now.

---

Assigning to @deontologician since he seems best able to review it (if that's alright?).

**Comment by marshall007** [comment]
@mlucy great, thanks!

**Comment by marshall007** [comment]
@mlucy oh one another thing was there were a couple weird discrepancies like `r.minval` produces `RqlInternalError` whereas `r.expr(r.minval).coerce_to('string')` produces `RqlLogicError` both with the same error message: `Cannot convert r.minval to JSON.`

**Comment by mlucy** [comment]
I think the first probably shouldn't be an internal error; I'll fix that.

Did you see any other weird ones like that?

**Comment by marshall007** [comment]
`table.index_create('id')` produced `RqlLogicError` instead of `RqlOpFailedError`, but after thinking about it that one probably makes sense. The only other one I remember was invalid optarg `array_limit: -1` producing `RqlInternalError` vs deprecated optarg `return_vals` producing `RqlLogicError`.

**Comment by deontologician** [comment]
This looks good to me, so I think we can merge this once the changes corresponding to #4484 are made

File: drivers/javascript/errors.coffee [hunk 1]
```diff
@@ -24,9 +24,19 @@ class RqlServerError extends Error
 class RqlRuntimeError extends RqlServerError
 
 class RqlCompileError extends RqlServerError
-
 class RqlClientError extends RqlServerError
 
+class RqlInternalError extends RqlRuntimeError
+class RqlResourceError extends RqlRuntimeError
+class RqlLogicError extends RqlRuntimeError
+class RqlNonExistenceError extends RqlRuntimeError
```
Comment (mlucy) [inline drivers/javascript/errors.coffee]: A `NonExistenceError` is a type of `LogicError` (currently it's the only type we report on its own), so it should probably extend `RqlLogicError` instead.

Comment (mlucy) [inline drivers/javascript/errors.coffee]: (This probably could be clearer in the protobuf, I tried to encode the hierarchy in the numbers used for the error types but it probably came out pretty opaque.)

Comment (marshall007) [inline drivers/javascript/errors.coffee]: @mlucy thanks, just pushed a fix.

**Comment by danielmewes** [comment]
Thinking about this again, I think we should merge this regardless of #4484 so we can include it in the 2.1 beta. If the tests need to be adapted, we can do that after #4484 gets in. (@deontologician)

Do the test changes currently in this pull request assume the current behavior or the one with #4484 fixed?

**Comment by marshall007** [comment]
@danielmewes test changes assume current behavior.

**Comment by danielmewes** [comment]
Great, then - assuming everything else is right - I think we should merge this.

**Comment by marshall007** [comment]
The only thing that hasn't been addressed here is the failing connections tests (quoting myself):

> Also there are a few failing connections tests in the JS driver right now. Two of them are timeouts, and the others are "could not connect to server" errors for r.http. So there may be a couple tests in there that still need to be updated, but I was unsure of why they were failing to begin with.

Part of that is probably #4496, but also some tests in `connections/http.js` weren't updated because I couldn't get them to run. I think it has something to do with httpbin being proxied and that wasn't working for me.

**Comment by deontologician** [comment]
OK, I have a patch that will fix the remaining things needed in connection/http.js, so I'm going to merge this and apply that directly. Thanks again for the PR @marshall007

**Pull request merged**